### ISSUE TITLE
./ in @INC leads to wrong results in %INC

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9777]
Migrated from [rt.perl.org#66942](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=66942) (status was 'resolved')

Searchable as RT66942$

**Comment by p5pRT** [comment]
### From frank.wiegand@gmail.com

#### Created by frank.wiegand@gmail.com



Unshifting the »\./« path to @&#8203;INC leads to wrong results in the %INC hash&#8203;:

&nbsp;  % pwd
&nbsp;  /tmp
&nbsp;  
&nbsp;  % cat Foo\.pm
&nbsp;  package Foo;
&nbsp;  1;
&nbsp;  
&nbsp;  % perl \-le 'unshift @&#8203;INC\, "\./"; require Foo; print $INC\{"Foo\.pm"\}'
&nbsp;  /Foo\.pm

Output should be »\./Foo\.pm«\, »/Foo\.pm« does not exist\.


Note&#8203;: This does not affect all dotty/slashy/relative paths&#8203;:

&nbsp;  % perl \-le 'unshift @&#8203;INC\, "\./\."; require Foo; print $INC\{"Foo\.pm"\}'
&nbsp;  \./Foo\.pm


Thanks to Christoph Lamprecht\, who presented this bug at yesterdays Berlin\.pm meeting\.

Bye\, Frank

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.10.0:

Configured by Debian Project at Fri Jun 12 21:43:19 UTC 2009.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.6.26-1-vserver-amd64, archname=x86_64-linux-gnu-thread-multi
    uname='linux excelsior 2.6.26-1-vserver-amd64 #1 smp sat jan 10 19:46:42 utc 2009 x86_64 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=x86_64-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.10 -Darchlib=/usr/lib/perl/5.10 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.10.0 -Dsitearch=/usr/local/lib/perl/5.10.0 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.10.0 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.3.3', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.9.so, so=so, useshrplib=true, libperl=libperl.so.5.10.0
    gnulibc_version='2.9'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.10.0:
    /etc/perl
    /usr/local/lib/perl/5.10.0
    /usr/local/share/perl/5.10.0
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.10
    /usr/share/perl/5.10
    /usr/local/lib/site_perl
    .


Environment for perl 5.10.0:
    HOME=/home/fw
    LANG=de_DE.UTF-8
    LANGUAGE=
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/sbin:/usr/sbin:/home/fw/bin:/home/fw/bin:/usr/local/bin:/usr/bin:/bin:/usr/games
    PERL_AUTOINSTALL=--defaultdeps
    PERL_BADLANG (unset)
    PERL_EXTUTILS_AUTOINSTALL=--defaultdeps
    PERL_MM_USE_DEFAULT=1
    SHELL=/bin/zsh


```
</details>

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From @rgs

2009/6/25 Frank Wiegand \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> Unshifting the »\./« path to @&#8203;INC leads to wrong results in the %INC hash&#8203;:
>
>  % pwd
>  /tmp
>
>  % cat Foo\.pm
>  package Foo;
>  1;
>
>  % perl \-le 'unshift @&#8203;INC\, "\./"; require Foo; print $INC\{"Foo\.pm"\}'
>  /Foo\.pm
>
> Output should be »\./Foo\.pm«\, »/Foo\.pm« does not exist\.

Thanks for the report\. This should be solved by the following change&#8203;:

commit e63be7465b2d564165a64209685cfd7026a2005b
Author&#8203;: Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;gmail\.com>
Date&#8203;:   Thu Jun 25 19&#8203;:57&#8203;:19 2009 \+0200

&nbsp;    Trim all trailing / from "\." in @&#8203;INC when filling %INC

&nbsp;    This fixes bug \#66942 : as a / was left in the directory name\,
&nbsp;    $INC\{"Foo\.pm"\} for a file loaded from the current directory
&nbsp;    was given the incorrect value "/Foo\.pm"\.

 pp\_ctl\.c       |    6 \+\+\+\+\-\-
 t/run/runenv\.t |   17 \+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\+\-
 2 files changed\, 20 insertions\(\+\)\, 3 deletions\(\-\)
